File: python/zpy/src/zingg_py/cli_labeller.py
"""Interactive CLI labeller — port of ``CliLabeller.scala``.

Prints each pair as a side-by-side table and reads one answer from stdin:
  y / 1 / yes → match (1.0)
  n / 0 / no  → non-match (0.0)
  s           → skip this pair (no label)
  q / quit    → quit; remaining pairs get no label
"""

from __future__ import annotations

import sys

import pandas as pd

from . import pair_table
from .labeller import Labeller, OnlineAnswer
from .schema import ZinggConf


class CliLabeller(Labeller):
    def __init__(self, out=sys.stdout, inp=sys.stdin, width: int | None = None):
        self._out = out
        self._in = inp
        self._width = width if width is not None else pair_table.default_width()

    def ask(self, row, cfg: ZinggConf, header: str) -> OnlineAnswer:
        print("", file=self._out)
        print(header, file=self._out)
        for line in pair_table.render(cfg, row, self._width):
            print(line, file=self._out)
        print("Match? [y]es / [n]o / [s]kip / [q]uit: ", end="", file=self._out, flush=True)
        line = self._in.readline()
        if line == "":  # EOF
            return OnlineAnswer.quit()
        ans = line.strip().lower()
        if ans in ("y", "1", "yes"):
            return OnlineAnswer.label(1.0)
        if ans in ("n", "0", "no"):
            return OnlineAnswer.label(0.0)
        if ans in ("q", "quit"):
            return OnlineAnswer.quit()
        return OnlineAnswer.skip()

    def label(self, pairs: pd.DataFrame, cfg: ZinggConf) -> pd.DataFrame:
        kept, labels = [], []
        total = len(pairs)
        for idx, (_, r) in enumerate(pairs.iterrows()):
            ans = self.ask(r, cfg, f"━━━ Pair {idx + 1} / {total} ━━━")
            if ans.is_quit:
                break
            if ans.is_label:
                kept.append(r)
                labels.append(ans.value)
        if not kept:
            out = pairs.iloc[0:0].copy()
            out[cfg.label_col] = pd.Series([], dtype=float)
            return out
        out = pd.DataFrame(kept).reset_index(drop=True)
        out[cfg.label_col] = labels
        return out

File: zscala/src/main/scala/zingg/zs/CliLabeller.scala
package zingg.zs

import org.apache.spark.sql.{DataFrame, Row}

import scala.io.StdIn

/** Interactive CLI labeller: prints each pair to stdout and reads y/n/s
  * from stdin.
  *
  *   y / 1 → match (1.0)
  *   n / 0 → non-match (0.0)
  *   s     → skip this pair (no label produced)
  *   q     → quit; remaining pairs get no label
  */
final class CliLabeller(
    out: java.io.PrintStream = System.out,
    in:  java.io.BufferedReader = new java.io.BufferedReader(new java.io.InputStreamReader(System.in)),
    width: Int = CliLabeller.defaultWidth
) extends Labeller {

  def label(pairs: DataFrame, cfg: ZinggConf): DataFrame = {
    import org.apache.spark.sql.types.{DoubleType, StructField, StructType}
    val rows = pairs.collect()
    var quit = false
    val outRows = rows.zipWithIndex.flatMap { case (r, idx) =>
      if (quit) None
      else askOne(r, idx, rows.length, cfg) match {
        case CliLabeller.Quit   => quit = true; None
        case CliLabeller.Skip   => None
        case CliLabeller.Lab(v) =>
          Some(Row.fromSeq((0 until r.length).map(r.get) :+ v))
      }
    }
    val schema = StructType(pairs.schema.fields :+
      StructField(cfg.labelCol, DoubleType, nullable = false))
    val spark = pairs.sparkSession
    spark.createDataFrame(spark.sparkContext.parallelize(outRows.toSeq, 2), schema)
  }

  private def askOne(r: Row, idx: Int, total: Int, cfg: ZinggConf): CliLabeller.Answer = {
    out.println()
    out.println(s"━━━ Pair ${idx + 1} / $total ━━━")
    PairTable.render(cfg, r, width).foreach(out.println)
    out.print("Match? [y]es / [n]o / [s]kip / [q]uit: ")
    out.flush()
    Option(in.readLine()).map(_.trim.toLowerCase).getOrElse("q") match {
      case "y" | "1" | "yes" => CliLabeller.Lab(1.0)
      case "n" | "0" | "no"  => CliLabeller.Lab(0.0)
      case "q" | "quit"      => CliLabeller.Quit
      case _                 => CliLabeller.Skip
    }
  }
}

object CliLabeller {
  /** Terminal width from `$COLUMNS` when sane, else a readable default. */
  def defaultWidth: Int =
    sys.env.get("COLUMNS").flatMap(s => scala.util.Try(s.trim.toInt).toOption)
      .filter(_ >= 40).getOrElse(100)

  sealed trait Answer
  case object Quit             extends Answer
  case object Skip             extends Answer
  final case class Lab(v: Double) extends Answer
}
